Allow selective exclusion of Lagrange from p-refinement

lindsayad opened the pull request:
Refs #25105

lindsayad committed:
Support selective p-refinement

lindsayad committed:
Restore original computeMarker routine

lindsayad committed:
Don't write constant monomial as elemental when there's p-refinement

lindsayad committed:
Use add_p_level arg when computing ICs

lindsayad committed:
Add selective refinement test

Closes #25105

lindsayad committed:
Update to new variable group based API

zachmprince commented on framework/src/ics/InitialConditionTempl.C:
```diff
@@ -173,11 +173,13 @@ InitialConditionTempl<T>::compute()
   // From here on out we won't be sampling at nodes anymore
   _current_node = nullptr;
 
+  const bool add_p_level = _var.dofMap().should_p_refine(_fe_type.family);
```
After much debugging, found this was a problem I was having when it was still trying to p-refine lagrange variables. I think this is proper thing to do:

```suggestion
const bool add_p_level = _var.dofMap().should_p_refine(_var.dofMap().var_group_from_var_number(_var.number()));
```

lindsayad replied:
Oh wow, implicit conversion led to me not updating that code. I will probably delete the implicit conversion to ensure that can’t happen in the future

lindsayad replied:
sorry you had to debug that!

lindsayad replied:
opened https://github.com/libMesh/libmesh/pull/3619

lindsayad replied:
ok fixed this line

lindsayad committed:
Initialize stateful material properties after initial mesh adaptivity

It seems very silly to initialize stateful material properties before.
By doing so you require projection operations. But I'm sure this
seemingly intelligent change will cause something to fail

lindsayad committed:
As in h-adaptivity, only execute elem when it is >= neighbor refinement

lindsayad commented:
> Initialize stateful material properties after initial mesh adaptivity
> It seems very silly to initialize stateful material properties before.
> By doing so you require projection operations. But I'm sure this
> seemingly intelligent change will cause something to fail

My commit message was so on point

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1759084/) on d132f27 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25108/site)

This comment will be updated on new commits.

lindsayad committed:
Add stateful property w/ p-refinement tests

lindsayad committed:
Revert "Initialize stateful material properties after initial mesh adaptivity"

This reverts commit dbdc27d9b48ac042a47bb780a119ec433d4bcd01.

lindsayad committed:
Doxygen/developer documentation of refinement maps in MooseMesh

lindsayad committed:
Move level (h or p) check down to ThreadedElementLoopBase

This necessitates moving havePRefinement data down into MooseMesh
since the base loop doesn't have access to a Problem

lindsayad committed:
Add p-refinement branches for writing refine/coarsen maps

lindsayad committed:
Typo fix in Adaptivity code comment

lindsayad committed:
Add absolute tolerance for spatial adaptivity test

moosebuild commented:
Job [Python 3.8](https://civet.inl.gov/job/1765045/) on a85a98e : invalidated by @lindsayad

lindsayad committed:
Don't add inactive just-p-refined elements to change list

lindsayad committed:
Assert elements active in FlagElementsThread

lindsayad committed:
Switch an if check so that we can do p-refinement after h-refinement

lindsayad committed:
get/setHavePRefinement -> doingPRefinement

lindsayad committed:
_have_p_refinement -> _doing_p_refinement

lindsayad committed:
Finish API changes for paradigm allowing alternate h/p-refinement

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1765239/) on 84c371c wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>1acda2</th>
      <th colspan=4>#25108 84c371</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.70%</td>
      <td>85.71%</td>
      <td>+0.01%</td>
      <td>97.73%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>91495</td>
      <td>91570</td>
      <td>+75</td>
      <td>215</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15266</td>
      <td>15263</td>
      <td>-3</td>
      <td>5</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25108/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25108/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25108/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Make a bunch of separate routines for p-refinement

I need to make sure I have maps for different levels of
p-refinement, e.g. not just 0 to 1, in order to get the
quadrature point mapping correct. And reusing the h-refinement
machinery makes that difficult to do

lindsayad committed:
Use different file base for stateful p tests

lindsayad committed:
Finally good results for stateful material props with p-refinement

- Had to make sure we had all the maps for different levels of
  p-refinement
- Had to make sure not to erase our p-refined elements from the
  material property storage
- Had to make sure that when writing into our stateful material
  property we cloned the property first because we are writing
  and reading from conceptually the same property

GiudGiud commented on framework/include/base/Adaptivity.h:
```diff
@@ -129,9 +129,11 @@ class Adaptivity : public ConsoleStreamInterface,
+   * @param disable_lagrange_p_refinement Whether to disable p-refinement of Lagrange variables
    */
-  void switchHToPRefinement();
+  void doingPRefinement(bool doing_p_refinement, bool disable_lagrange_p_refinement);
```
what s the hate against Lagrange p-refinement

GiudGiud replied:
ok it s against nodal

@roystgnr am I correct in that the problem (needing a second order mesh to do p-refinement) is the same for:
HERMITE,  CLOUGH in general (some dofs at nodes, some gradient dofs)
HIERARCHIC, SZABAB, BERNSTEIN, RATIONAL_BERNSTEIN at some orders have dofs on nodes

roystgnr replied:
It's not directly a matter of p refinement, it's a matter of the combination of p with FEFamily.  As soon as you get a DoF that has support on a set of elements, however you got that DoF, you need a node which is connected to precisely that set of elements.  And even a second order mesh may not be enough - a TET10 will do p=1 or p=2 fine, but get up to p=3 on a C0 element and now you've got face degrees of freedom and you need a TET14 to give you face nodes to put them on.  That's true whether you got p=3 by declaring it from the get-go or by p-refining a p=2 variable once or a p=1 variable twice.  Every C0 or C1 variable type will get to the point where it needs edge and face nodes at *some* degree, depending on the element type; only the completely discontinuous bases don't ever need them.

Lagrange is extra tricky because since it's nodal rather than hierarchic we can't just get away with having e.g. one face node, we have to have one face node at exactly the spots where we have one Lagrange DoF evaluation for that p.  So we can support p=1 Lagrange on any element or p=2 on a quadratic (or in the future a quartic or any p=2*n element), but we *can't* support p=2 Lagrange on a cubic element.  A Hex64 will have 4 face nodes, and will work fine with a cubic Lagrange, but none of those nodes are in the center of the face where quadratic Lagrange needs a node.  So unless you "skip" p levels, p refinement on Lagrange can go from 1 to 2 and then it gets stuck and that's just sad.

Rational_Bernstein (the reason why I'll be adding Hex64 and other higher order Hex/Quad/Edge elements)  is kind of an in-between case - the Bernstein DoFs aren't point evaluations on nodes, but they're associated with nodes and the Rational<Foo> weightings get stored on nodes.  IGA meshes will have some of the same restrictions as Lagrange for that reason.

GiudGiud replied:
so it's a matter of which family uses the additional second / third etc order nodes to start dofs. 
Just Lagrange and Rational Bernstein?

roystgnr replied:
`LAGRANGE_VEC` is also nodal (albeit multiple components per node).  I think that's the only other that can't be made to work easily with uniform p refinement.

With *adaptive* p refinement, though, we have the restriction that libMesh currently only constructs constraints for hierarchic bases (where the p=n+1 basis is just a superset of the p=n basis).  Discontinuous FE don't need constraints, but Bernstein and Clough need constraints that we can't construct for them.

Oh, and there's the fact that not everything that can support higher p in theory has the code written in practice.  `FEInterface::max_order` lists a bunch of not-implemented cases still: Szabo-Babuska is only hard-coded up to p=7, Clough for p=3, Nedelec-One and Raviart-Thomas for p=1.  We could add higher order capability to any of those but nobody has.

GiudGiud replied:
ok for Lagrange_vec being also purely nodal.

So both Szabab and Clough have DOFs at nodes and max allowed degrees that can be considered for uniform or adaptive p-refinement. But are they using the edge nodes of a TET10 and/or the face nodes of a TET14 in a way that precludes from hosting a third order on a TET10 / a second order on a TET14  ?

> It's not directly a matter of p refinement, it's a matter of the combination of p with FEFamily. 

this is what I'm trying to get to. Is it really just Lagrange variables that are problematic, or any variable in which we store DOFs at nodes will be problematic And if the latter, should Alex just replace lagrange to 'NodalDofed' to make it work in the text

lindsayad replied:
I could make this a parameter for which users can supply a list of families to disable, or variable groups but I think that families would be more user friendly. The only thing I really care about is getting this into users' hands sooner than later.

roystgnr replied:
> or any variable in which we store DOFs at nodes will be problematic

Any variable in which we store DoFs at nodes *can* be problematic, but doesn't have to be on the right mesh.  Do `all_complete_order()` (or load a mesh from file where all the Tets are TET14 etc., whatever), and anything for which `FE::is_hierarchic()` will be fine to p-refine up to its max order.

lindsayad replied:
I could change this to `disable_non_hierarchic`?

roystgnr replied:
That seems like a solid improvement to me.

Looking in libMesh, all our safety checking is in devel mode (because it necessarily involves a test at every element/neighbor interface with a p mismatch), and that's not good enough for MOOSE users, so better to disable p-refinement of non-hierarchic elements entirely.  We could selectively re-enable it if people really want to use one of them for supported *uniform* p refinement.

GiudGiud replied:
I think Roy is saying this is only OK for all_complete_order meshes.

so it s ok if you assert after that the mesh is as expected?

lindsayad replied:
> so it s ok if you assert after that the mesh is as expected?

I don't know what "that" means, or what I'm supposed to be asserting.

It's certainly valid for a user running with DFEM with monomial variables (a large user base) to run with a mesh that is not `all_complete_order()`

GiudGiud replied:
well monomial is fine for this problem of nodal values right
I think the concern is has_nodal_dofs_non_hierarchic variables

roystgnr replied:
FWIW we'll already be hitting decent error messages if we overrefine an FE without all the right Elem types to support it; the same one we hit now if someone tries to use e.g. second order Lagrange on a HEX8.

roystgnr replied:
Though I admit an error before refining would be more user friendly than an error after.

lindsayad replied:
@GiudGiud sure, but I don't understand what you're proposing we do about it. Re-implement the exact same error messages as already exist in libMesh but just do it before we p-refine? I don't see how much we gain there. Generally, we cannot know how much p-refinement there will be *a priori* at the start of a simulation, so I don't know how much user time you will save by giving an error message right before p-refinement as opposed to right after. 

I think what does make sense is to allow the user to disable p-refinement of different finite element families. Perhaps as I suggested before this parameter can be changed to be a `MultiMooseEnum` and the user can specify what families they want to disable because they know that they might or will hit a p-refinement error and they don't care that a given variable have the p-order applied to it. Or if we really want to grant flexibility it can be a `std::vector<MultiMooseEnum>` in which the user can specify disabling of p-refinement for FEFamilies of given systems.

GiudGiud replied:
Ok I see your point. If the errors already exist in libmesh, we're already covered.
It seems fitting to make it a MultiMooseEnum, so we catch both lagrange and lagrange vec from the get go

GiudGiud commented on framework/include/base/Assembly.h:
```diff
@@ -2755,6 +2757,9 @@ class Assembly
   /// storing the dof indices
   std::vector<dof_id_type> _row_indices, _column_indices;
+
+  /// Whether we have adjusted our FE objects due to the presence of p-refinement
```
specify what you mean by adjusted

GiudGiud commented on framework/include/base/Assembly.h:
```diff
@@ -1834,8 +1834,10 @@ class Assembly
   /**
    * Indicate that we have p-refinement
+   * @param disable_lagrange_p_refinement Indicate whether we should disable p-refinement of
+   * Lagrange objects
```
list the types of objects concerned

GiudGiud commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -90,6 +90,14 @@ class MaterialPropertyStorage
                             const int input_child,
                             const int input_child_side);
 
+  void updateStatefulPropsP(const processor_id_type pid,
```
docstring

GiudGiud replied:
also can we find a word instead of "P"

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1268,6 +1268,23 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
+   */
+  [[nodiscard]] bool doingPRefinement() const { return _doing_p_refinement; }
+
+  const std::vector<QpMap> & getPRefinementMap(const Elem & elem) const;
```
add a (shared  if you want) docstring

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1575,6 +1592,15 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
                             int child,
                             int child_side);
 
+  void buildHRefinementAndCoarseningMaps(Assembly * assembly);
```
I ll look at the implementation. Not sure whether using assembly directly in the mesh is where we want to be.
Maybe a util is warranted

GiudGiud commented on framework/include/mesh/MooseMesh.h:
```diff
@@ -1575,6 +1592,15 @@ class MooseMesh : public MooseObject, public Restartable, public PerfGraphInterf
 
+  void buildHRefinementAndCoarseningMaps(Assembly * assembly);
+
+  const std::vector<QpMap> & getPRefinementMapHelper(
```
I dont understand yet why it s the mesh's responsibility to handle p-refinement. Seems like a FE thing to me

lindsayad replied:
`p-refinement` is a flag on the `Elem`. It is treated analogously to h-refinement

GiudGiud commented on framework/src/actions/SetAdaptivityOptionsAction.C:
```diff
@@ -50,6 +44,25 @@ SetAdaptivityOptionsAction::validParams()
+      "initial_steps", 0, "The number of adaptive steps to do based on the initial condition.");
+  params.addParam<MarkerName>(
+      "initial_marker",
+      "The name of the Marker to use to adapt the mesh during initial refinement.");
```
param group for initial adaptivity

GiudGiud commented on framework/src/base/Assembly.C:
```diff
@@ -4745,42 +4754,55 @@ Assembly::havePRefinement()
+        {
+          auto & fe_container_dim = libmesh_map_find(fe_container, dim);
+          auto fe_it = fe_container_dim.find(lagrange_type);
+          if (lagrange_type == helper_type)
+            mooseAssert(fe_it != fe_container_dim.end(), "We should have the helper type");
```
never seen that

how does it compile in opt? do mooseAssert completely disapeear (thus problem), or do they turn into a neutral statement

lindsayad replied:
```
#ifdef NDEBUG
#define mooseAssert(asserted, msg) ((void)0)
#else
```

GiudGiud replied:
thx

GiudGiud commented on framework/src/base/Assembly.C:
```diff
@@ -4721,15 +4722,23 @@ Assembly::helpersRequestData()
-                                      const bool user_added_helper_type,
-                                      auto & fe_container)
+  std::vector<FEType> lagrange_types = {helper_type};
+  if (helper_order == SECOND)
```
and if user_disabled_lagrange_p_refinement ?

lindsayad replied:
Then the `helper_order` is still `SECOND`. The helper order is setup at Assembly construction time

GiudGiud commented on framework/src/outputs/AdvancedOutput.C:
```diff
@@ -429,13 +429,19 @@ AdvancedOutput::initAvailableLists()
-        if (type.order == CONSTANT && type.family != MONOMIAL_VEC)
+        // A note that if we have p-refinement we assume "worst-case" scenario that our constant
+        // monomial/monomial-vec families have been refined and we can no longer write them as
+        // elemental
```
this is making me think of transfers

both nearest-location and shape evaluation have special logic for constant monomials

we need to be returning the current order of the family, not the initial one, to avoid fooling downstream code

GiudGiud replied:
whole auxkernel system has a special treatment for const monomial iirc

lindsayad replied:
I've opened https://github.com/idaholab/moose/issues/25465. The aux kernel system checks the number of dofs so it's robust

GiudGiud replied:
sounds good

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -7140,13 +7143,16 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
+      // not thread safe in and of itself because it is a read-write operation. Note that we do not
+      // do the erasure for p-refinement because the coarse level element is the same as our active
+      // refined level element
+      if (!doingPRefinement())
```
so it s either p or h refinement right
there is no mixing allowed

lindsayad replied:
At the moment for a given `Adaptivity::adaptMesh` you must be either h- or p-, not both. In terms of the code APIs you could `adaptMesh` with h-, and then `adaptMesh` with p, and then `adaptMesh` again with h, so on and so forth. 

In terms of an input file, though, right now you cannot mix. There is only the `switch_h_to_p_refinement` flag in the Adaptivity action that is a one-time toggle. You *can* do initial uniform h-refinement and then later p-refinement based on the `Marker` system, which is demo'd in a couple of the tests.

GiudGiud commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -63,6 +63,53 @@ MaterialPropertyStorage::eraseProperty(const Elem * elem)
+
+  getMaterialData(tid).resize(n_qpoints);
+
+  mooseAssert(elem.active(), "We should be doing p-refinement");
```
```suggestion
  mooseAssert(elem.active(), "We should be doing p-refinement on active elements only");
```

GiudGiud commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -63,6 +63,53 @@ MaterialPropertyStorage::eraseProperty(const Elem * elem)
+      // We need to clone this property in order to not overwrite the values we're going to be
+      // reading from
+      auto previous_p_level_prop = current_p_level_prop.clone(current_p_level_prop.size());
+      // Cloning, despite its name, does not copy the data. Luckily since we are about to overwrite
```
"despite its name" lol

GiudGiud commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -63,6 +63,53 @@ MaterialPropertyStorage::eraseProperty(const Elem * elem)
+      // all of the current_p_level_prop data, we can just swap its data over to our
+      // previous_p_level_prop
+      previous_p_level_prop->swap(current_p_level_prop);
+      current_p_level_prop.resize(n_qpoints);
```
why is it a different size? are we increasing the qrule size during p refinement?

lindsayad replied:
Yes, the p-level order is added to the attached quadrature order. This allows you to automatically continue to exactly integrate quantities a user expects to be exactly integrable, etc.

GiudGiud replied:
makes sense

GiudGiud reviewed (commented):
lots of questions

reviewed everything but moosemesh.C

lindsayad committed:
Use self comm for mesh for p-refinement map

Otherwise I think there are big conceptual issues. E.g. if I am
distributed I may be adding a different element on each process,
which definitely isn't a ReplicatedMesh

lindsayad committed:
Address Guillaume review:

- Clarifying comments around Assembly::havePRefinement
- updateStatefulPropsP -> updateStatefulPropsForPRefinement
- Doxygen
- Initial adaptivity param grouping
- Clarify assertion comment

lindsayad committed:
Allow arbitrary disablement of FEFamilies

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1796745/) on ef6aa66 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25108/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25108/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format dcbdc8afa1fcfd6f272fc0af31dc6807ccd8c633`

moosebuild commented:
Job [Parallel sweep evens](https://civet.inl.gov/job/1797758/) on f402d67 : invalidated by @lindsayad

lindsayad commented:
This is ready for further review

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2070,6 +2070,114 @@ MooseMesh::buildRefinementAndCoarseningMaps(Assembly * assembly)
+    auto & qrule = assembly->writeableQRule();
+    auto & qrule_face = assembly->writeableQRuleFace();
+
+    libMesh::Parallel::Communicator self_comm{};
```
add comments
seems like you re using assmelby to get the missing nodes to support p-refinement and then adding them to the mesh

lindsayad replied:
will get these next time around

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2070,6 +2070,114 @@ MooseMesh::buildRefinementAndCoarseningMaps(Assembly * assembly)
+    // Need to do this just once to get the right qrules put in place
+    assembly->setCurrentSubdomainID(moose_elem->subdomain_id());
+    assembly->reinit(moose_elem);
+    assembly->reinit(moose_elem, 0);
```
```suggestion
    assembly->reinit(moose_elem, /*side_index=*/0);
```

GiudGiud commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2070,6 +2070,114 @@ MooseMesh::buildRefinementAndCoarseningMaps(Assembly * assembly)
+    fe_face->attach_quadrature_rule(qrule_face);
+    fe->reinit(elem);
+    volume_ref_points_coarse = qrule->get_points();
+    fe_face->reinit(elem, (unsigned int)0);
```
same

GiudGiud reviewed (changes requested):
Did some testing. 
With second order elements (in the basic test), and with exotic FE families (bernstein, hermite, hierarchic, with first/second order elements) I hit asserts in devel and get past them in opt

user docs should probably warn against starting with higher order elements and using exotic families if not fixed

lindsayad commented:
> With second order elements (in the basic test)

@roystgnr get ready for a lot more `add_p_level`! 😆

lindsayad committed:
Add more tests

- first order lagrange on second order mesh
- third order hermite on first order mesh
- third order hermite on second order mesh

lindsayad committed:
Fix FE lower instantiations for vector vars

lindsayad committed:
Allow disabling p-refinement of vector FE types

lindsayad committed:
Add hierarchic and vector FE tests

lindsayad commented:
Added a lot more tests. Will require https://github.com/libMesh/libmesh/pull/3652

roystgnr commented on framework/src/base/Assembly.C:
```diff
@@ -402,7 +402,11 @@ Assembly::buildVectorLowerDFE(FEType type) const
   // the current mesh minus one (because this is for lower-dimensional
   // elements!)
-  for (unsigned int dim = 0; dim <= _mesh_dimension - 1; dim++)
+  unsigned int dim = ((type.family == LAGRANGE_VEC) || (type.family == MONOMIAL_VEC)) ? 0 : 2;
```
Is the intent here to allow all other vector-valued FE types?  Or should we test for `FEInterface::field_type(type.family) != TYPE_SCALAR` instead of hard-coding?

lindsayad replied:
> Is the intent here to allow all other vector-valued FE types?

Yes that is the intent. The choice to allow dim < 2 FEs for LAGRANGE_VEC and MONOMIAL_VEC is somewhat arbitrary and could be removed; I don't think anyone has a use for it. But this most recent code was based of looking at `build` in `FEGenericBase`. Since we allow building those dimensions in libMesh I figured I would allow it here

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/1872980/) on f1299bd : invalidated by @lindsayad

lindsayad commented:
libMesh update is in, so this PR is ready for review once more

GiudGiud commented:
Looks good except on this
```
 ../../../moose_test-opt -i test.i  AuxVariables/test/family=LAGRANGE_VEC Mesh/second_order=true
```

```
Memory leak detected!
Compile in DEBUG mode with --enable-reference-counting
for more information
```
who is throwing that?

GiudGiud commented:
umm nevermind it looks like it happens without adaptivity

GiudGiud reviewed (approved):
issues are unrelated

lindsayad committed:
dont-p-refine-lagrange -> dont-p-refine

lindsayad committed:
Fix memory leak for lower-D vector FEs

lindsayad committed:
Disable p-refinement of continuous non-hierarchic bases by default

And Nedelec as well, and document this.

roystgnr commented:
I've got a couple questions; reviewing now.

roystgnr commented on framework/src/loops/ProjectMaterialProperties.C:
```diff
@@ -156,6 +212,8 @@ ProjectMaterialProperties::onInternalSide(const Elem * elem, unsigned int /*side
   if (_need_internal_side_material &&
       _refine) // If we're refining then we need to also project "internal" child sides.
   {
+    mooseError("I'm pretty sure we're not handling stateful material property pronlongation or "
```
```suggestion
    mooseError("I'm pretty sure we're not handling stateful material property prolongation or "
```

roystgnr reviewed (approved):
No remaining concerns that can't be fixed later, I think.

lindsayad commented on framework/include/loops/ThreadedElementLoopBase.h:
```diff
@@ -404,16 +404,24 @@ bool
-  return (neighbor.active() && (neighbor.level() == elem.level()) && (elem_id < neighbor_id)) ||
-         (neighbor.level() < elem.level());
+  return (neighbor.active() && (neighbor_level == elem_level) && (elem_id < neighbor_id)) ||
+         (neighbor_level < elem_level);
```
This PR has already ballooned way beyond what I had initially planned for. I've opened https://github.com/idaholab/moose/issues/26047 which we can refer to when giving first class support for alternating h- and p-adapvity

lindsayad commented on framework/src/actions/SetAdaptivityOptionsAction.C:
```diff
@@ -139,6 +156,14 @@ SetAdaptivityOptionsAction::act()
+      if (!isParamSetByUser("disable_p_refinement_for_families"))
+        // If the user has not set this parameter we will set a logicial default
+        disable_p_refinement_for_families =
+            "LAGRANGE NEDELEC_ONE LAGRANGE_VEC CLOUGH BERNSTEIN RATIONAL_BERNSTEIN";
```
I agree with you. I did not want to introduce `RAVIART_THOMAS` since it does not yet exist as one of our MOOSE available families. I believe https://github.com/idaholab/moose/pull/25838 is a stepping stone for that. I will open a libMesh issue for getting that `is_hierarchic` shim

lindsayad replied:
opened https://github.com/libMesh/libmesh/issues/3705

lindsayad committed:
Typo fix for prolongation

Co-authored-by: roystgnr <roy@stogners.org>
